### ISSUE TITLE
Different validation scores before and after save on the same .pt

### ISSUE THREAD

**Issue by mittski** [mlfoundations/open_clip#738]
![Screenshot 2023-11-10 203014](https://github.com/mlfoundations/open_clip/assets/120546600/f5636333-4ab6-4147-9a54-5aba6d44cbb4)
_Val loss for Run & resume 1: Orange & Blue / Run & resume 2, gradscalar forced: pink & Yellow_

Reattempting to validate the same .pt previously saved shows a deterioration across all validation metrics; despite being untouched. (This second val assessment was performed with a config that performed no training steps; it should be reading the same .pt that was written last save without any adulteration, and of course the val set had no changes made)

**Edit:** I first thought it was a case of torch.load/save down out of a higher precision on save, but that doesn't seem to be the case

**Edit2:** I then checked if it was due to the conditional logic of setting `Scaler = GradScaler()` only when `args.precision == "amp"`, which excludes ` "amp_bf16", "amp_bfloat16"` presets. I forced the gradscaler assignment `if "amp" in args.precision`, which would then let the .pt be saved with a GradScaler instead of `None` like before... but as Run 2 shows; it's made no difference

Anyone have any ideas?

**Comment by mittski** [comment]
The issue is resolved: If you're resuming a .pt that was finetuned from an OpenAI model, it still needs the `pretrained="openai"` flag set on it in addition to the `resume="logs/xyz.pt"` for it to be re-loaded in correctly

Additional notes: changing your batch size between runs may affect your loss graph: but you'll know you've resumed wrong if your rank1/5/etc validation graphs are blasted

**Comment by Adenialzz** [comment]
Hi, thanks for sharing the solution. Have you found the reason for this error in the source code?

**Comment by mittski** [comment]
> Hi, thanks for sharing the solution. Have you found the reason for this error in the source code?

@Adenialzz The code goes through a bunch of initialization steps particular to whatever model you choose using `--pretrained` by checking an internal registry of models and doing a bunch of steps relevant to them; but `--resume` only loads the weights blindly. `--pretrained` will also try to blindly load weights if you feed it a filepath instead of a flag that matches the internal model registry

**Comment by Adenialzz** [comment]
In my case, `--pretrained` gives different results with `--resume` when I set them as a local file path. Do you have any idea what causes this difference? I find that it will also load state dict when `--pretrained` is a filepath, but I did not find the difference.
